nautobot/utilities/filters.py

```diff
@@ -286,6 +286,141 @@ def filter(self, qs, value):
         return qs
 
 
+class MappedPredicatesFilterMixin:
+    """
+    A filter mixin to provide the ability to specify fields and lookup expressions to use for
+    filtering.
+
+    A mapping of filter predicates (field_name: lookup_expr) must be provided to the filter when
+    declared on a filterset. This mapping is used to construct a `Q` query to filter based on the
+    provided predicates.
+
+    Example:
+
+        q = SearchFilter(
+            filter_predicates={
+                "comments": "icontains",
+                "name": "icontains",
+                "id": "iexact",
+            },
+        )
+
+    Optionally you may also provide a callable to use as a preprocessor for the filter predicate by
+    providing the value as a nested dict with "lookup_expr" and "preprocessor" keys. For example:
+
+        q = SearchFilter(
+            filter_predicates={
+                "asn": {
+                    "lookup_expr": "exact",
+                    "preprocessor": int,
+            },
+        )
+
+    This tells the filter to try to cast `asn` to an `int`. If it fails, this predicate will be
+    skipped.
+    """
+
+    # Optional label for the form element generated for this filter
+    label = None
+
+    def __init__(self, filter_predicates=None, label=None, *args, **kwargs):
+        if not isinstance(filter_predicates, dict):
+            raise TypeError("filter_predicates must be a dict")
+
+        # Format: {field_name: lookup_expr, ...}
+        self.filter_predicates = filter_predicates
+
+        # Try to use the label from the class if it is defined.
+        if label is None:
+            kwargs.setdefault("label", self.label)
+
+        super().__init__(*args, **kwargs)
+
+        # Generate the query with a sentinel value to validate it and surface parse errors.
+        self.generate_query(self.filter_predicates, value="")
+
+    def generate_query(self, filter_predicates, value):
```

Comment:
I was thinking ahead a little for both testability and introspection in the future since this can be used to deterministically generate a `Q` object based on predicates and value. Think: DynamicGroup reverse query generation based on saved filters. calling this directly would be more useful and testable.